File: packages/nakdimon-js/src/encode.ts
// Numeric encode/decode between the text layer (HebrewItem / normalized
// chars) and the model's integer vocabularies, per SPEC.inventories.
import { SPEC } from "./spec.gen.js";
import type { HebrewItem } from "./hebrew.js";

export type InventoryName = keyof typeof SPEC.inventories;

interface EncodedItems {
  letters: number[];
  niqqud: number[];
  dagesh: number[];
  sin: number[];
}

/**
 * Builds an encode map by iterating a decode array in order and letting
 * later duplicates overwrite earlier ones. This mirrors v1's
 * dict-overwrite CharacterTable and, for the v1 niqqud table, reproduces
 * the historic patah->15 encode_override for free (patah decodes from
 * index 9 but always encodes to 15, since 15 is written last).
 */
function buildEncodeMap(decode: readonly string[]): Map<string, number> {
  const map = new Map<string, number>();
  decode.forEach((c, i) => map.set(c, i));
  return map;
}

function encodeOne(map: Map<string, number>, c: string, kind: string): number {
  const id = map.get(c);
  if (id === undefined) {
    throw new Error(`unknown ${kind} character: ${JSON.stringify(c)} (U+${(c.codePointAt(0) ?? 0).toString(16)})`);
  }
  return id;
}

export class Inventory {
  readonly name: InventoryName;
  readonly letters: readonly string[];
  readonly niqqud: readonly string[];
  readonly dagesh: readonly string[];
  readonly sin: readonly string[];

  private readonly lettersEncode: Map<string, number>;
  private readonly niqqudEncode: Map<string, number>;
  private readonly dageshEncode: Map<string, number>;
  private readonly sinEncode: Map<string, number>;

  constructor(name: InventoryName) {
    const inv = SPEC.inventories[name];
    this.name = name;
    this.letters = inv.letters;
    this.niqqud = inv.niqqud.decode;
    this.dagesh = inv.dagesh.decode;
    this.sin = inv.sin.decode;

    this.lettersEncode = buildEncodeMap(this.letters);
    this.niqqudEncode = buildEncodeMap(this.niqqud);
    this.dageshEncode = buildEncodeMap(this.dagesh);
    this.sinEncode = buildEncodeMap(this.sin);
  }

  encodeLetters(normalized: string): number[] {
    return [...normalized].map((c) => encodeOne(this.lettersEncode, c, "letter"));
  }

  encodeItems(items: readonly HebrewItem[]): EncodedItems {
    return {
      letters: items.map((it) => encodeOne(this.lettersEncode, it.normalized, "letter")),
      niqqud: items.map((it) => encodeOne(this.niqqudEncode, it.niqqud, "niqqud")),
      dagesh: items.map((it) => encodeOne(this.dageshEncode, it.dagesh, "dagesh")),
      sin: items.map((it) => encodeOne(this.sinEncode, it.sin, "sin")),
    };
  }

  decodeNiqqud(i: number): string {
    return this.niqqud[i];
  }

  decodeDagesh(i: number): string {
    return this.dagesh[i];
  }

  decodeSin(i: number): string {
    return this.sin[i];
  }
}

File: packages/nakdimon/src/nakdimon/encode.py
"""Vocabulary <-> integer-id codec for the letters/niqqud/dagesh/sin channels,
built from spec["inventories"][name] (either "v1" or "v2")."""

from __future__ import annotations

from typing import Any

from nakdimon.hebrew import HebrewItem


class Inventory:
    def __init__(self, spec_inventory: dict[str, Any]) -> None:
        self.letters: list[str] = spec_inventory["letters"]
        self.letters_encode: dict[str, int] = {c: i for i, c in enumerate(self.letters)}

        self._niqqud_decode: list[str] = spec_inventory["niqqud"]["decode"]
        self._dagesh_decode: list[str] = spec_inventory["dagesh"]["decode"]
        self._sin_decode: list[str] = spec_inventory["sin"]["decode"]

        # Encoding by iterating decode-in-order and overwriting duplicates makes later
        # entries win — e.g. v1's patah appears at decode[9] and decode[15]; encoding
        # always produces 15 (the historic model-compatibility quirk spec.encode_overrides
        # documents). Assert against the spec's overrides so a decode-array edit that
        # silently changes this ordering fails loudly instead of drifting from the model.
        self.niqqud_encode: dict[str, int] = {c: i for i, c in enumerate(self._niqqud_decode)}
        for char, expected_id in spec_inventory["niqqud"].get("encode_overrides", {}).items():
            assert self.niqqud_encode[char] == expected_id, (
                f"niqqud encode_overrides mismatch: {char!r} encodes to "
                f"{self.niqqud_encode[char]}, expected {expected_id}"
            )

        self.dagesh_encode: dict[str, int] = {c: i for i, c in enumerate(self._dagesh_decode)}
        self.sin_encode: dict[str, int] = {c: i for i, c in enumerate(self._sin_decode)}

    def encode_letters(self, normalized: str) -> list[int]:
        return [self.letters_encode[c] for c in normalized]

    def encode_items(self, items: list[HebrewItem]) -> dict[str, list[int]]:
        return {
            "letters": [self.letters_encode[item.normalized] for item in items],
            "niqqud": [self.niqqud_encode[item.niqqud] for item in items],
            "dagesh": [self.dagesh_encode[item.dagesh] for item in items],
            "sin": [self.sin_encode[item.sin] for item in items],
        }

    def decode_niqqud(self, i: int) -> str:
        return self._niqqud_decode[i]

    def decode_dagesh(self, i: int) -> str:
        return self._dagesh_decode[i]

    def decode_sin(self, i: int) -> str:
        return self._sin_decode[i]
